src/main/java/io/cdap/delta/bigquery/BigQueryEventConsumer.java

```diff
@@ -458,26 +465,29 @@ private void loadStagingTable(TableId stagingTableId, TableBlob blob, int attemp
   }
 
   private void mergeStagingTable(TableId stagingTableId, TableBlob blob,
-                                 int attemptNumber) throws InterruptedException {
+                                 int attemptNumber) throws InterruptedException, DeltaFailureException {
     LOG.debug("Merging batch {} for {}.{}", blob.getBatchId(), blob.getDataset(), blob.getTable());
     TableId targetTableId = TableId.of(project, blob.getDataset(), blob.getTable());
-    Table targetTable = bigQuery.getTable(targetTableId);
-    if (targetTable == null) {
-      // We require that the table already exists and that it was created by this plugin when
-      // handling a DDL create table event.
-      // this is because the primary key for the table is contained in the DDL event, but not in DML events.
-      // so the primary key info is stored in the table definition.
-      // If we had the primary key information here, we could create the table now if it didn't exist for some reason
-      throw new IllegalStateException(
-        String.format("No target table %s.%s found. Please ensure that tables created by the replicator are not "
-                        + "deleted by other processes.", blob.getDataset(), blob.getTable()));
-    }
+    List<String> primaryKey = primaryKeyStore.get(targetTableId);
+    if (primaryKey == null) {
+      byte[] pkBytes;
+      try {
+        pkBytes = context.getState(String.format("%s-%s-primary-key", targetTableId.getDataset(),
+                                                 targetTableId.getTable()));
+      } catch (Exception e) {
+        throw new DeltaFailureException(
+          String.format("Not able to read state of primary key for table %s in dataset %s.",
+                        targetTableId.getTable(), targetTableId.getDataset()), e);
+      }
 
-    // TODO: figure app should provide a way to get table info given an offset
-    String pkStr = targetTable.getDescription();
-    pkStr = pkStr == null || !pkStr.startsWith("Primary Key: ") ?
-      "_sequence_num" : pkStr.substring("Primary Key: ".length());
-    List<String> primaryKey = Arrays.stream(pkStr.split(",")).collect(Collectors.toList());
+      if (pkBytes == null) {
+        throw new DeltaFailureException(
```

Comment:
this should never happen, unless there was some sort of manual deletion of the state. Let's change the message to:

Primary key information for table '%s' in dataset '%s' could not be found. This can only happen if state was manually deleted. Please create a new replicator and start again.

Reply:
Sure. I will update the message body.